### ISSUE TITLE
Destructuring with default value incorrectly throws an error

### ISSUE THREAD

**Issue by jondlm** [swc-project/swc#8527]
### Describe the bug

SWC is incorrectly handling a spread destructure with a default. I did some bisecting and found the bug was likely introduced by #6412 in v1.3.17.

[→ codepen](https://codepen.io/jondlm/pen/bGZWJyJ?editors=1111) comparing different versions of swc, babel, and raw browser eval

### Input code

```typescript
function foo({ a: { ...b } = {} }) {
  console.log(b);
}

foo({}); // should log `{}` but throws `Uncaught TypeError: Cannot destructure undefined`
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": true
    },
    "transform": {
      "react": {
        "refresh": true
      }
    },
    "target": "es2015",
    "minify": {
      "mangle": false,
      "compress": false,
    },
    "loose": true,
  },
  "module": {
    "type": "es6"
  },
  "isModule": true,
  "minify": false
 }
```

### Playground link (or link to the minimal reproduction)

https://play.swc.rs/?version=1.3.100&code=H4sIAAAAAAAAAx3LMQ6AIAwF0L2n%2BCMsZZd4FwRFTQgdkKnp3TXOL6%2FOXp5bOqqIU2wLFMycYVihBvNQAor0Ie3gJqfLPpIR%2FcF8RAgYl8y241MktUQvpvbS31YAAAA%3D&config=H4sIAAAAAAAAA1WQSw7CMAxE95yi8poFIMGCO3CIKDglqPnITiWqirvjpE1Jd%2FEbe8bOfOg6eLOGezfLU4qoiJG2WghPPqmPEEhTRNZkY4JjVRNnKdGIhXwXARIpzyaQa50IlU4NKMgQ8qu1EJO9laIe8xggX07n65oNznprptbfKd8PKMSogXFbUQcXJYT3QrUfQmBc8zMqHFx4jsVq%2FZZ8%2BrLCDbYmy4%2FaVqf%2FW5Wog9zyA0NXBE5kAQAA

### SWC Info output

    Operating System:
        Platform: darwin
        Arch: arm64
        Machine Type: arm64
        Version: Darwin Kernel Version 23.2.0: Wed Nov 15 21:53:18 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6000
        CPU: (10 cores)
            Models: Apple M1 Max

    Binaries:
        Node: 18.14.0
        npm: 9.3.1
        Yarn: 1.22.19
        pnpm: 8.6.0

    Relevant Packages:
        @swc/core: 1.3.100
        @swc/helpers: N/A
        @swc/types: N/A

    SWC Config:
        output: N/A
        .swcrc path: N/A

    Next.js info:
        output: N/A

### Expected behavior

The compiled code should execute and log `{}` without an exception.

### Actual behavior

It throws `Uncaught TypeError: Cannot destructure undefined`.

### Version

1.3.100

**Comment by magic-akari** [comment]
What's your Babel config?

I attempted to use `Chrome 59`, and encountered the same error in Babel.

https://babeljs.io/repl#?browsers=chrome%2059&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=GYVwdgxgLglg9mABMOcAUBvRBDAXIrAOmICNEBfRAXgMvIEoCAoRRCBAZzgBsBTQ7nADmaEvQDcTckyYp0GBuMQB6ZYg4ALOCG4ATRIKGIABguNMgA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact%2Cstage-2&prettier=false&targets=&version=7.23.8&externalPlugins=&assumptions=%7B%7D

**Comment by jondlm** [comment]
Interesting. It looks like the problem starts showing up in Babel with `chrome 51` but it's okay on `chrome 50`. Once we get to chrome 60 the problem goes away since chrome started supporting object rest and spread natively. So it seems like a babel bug as well.

**Comment by Austaras** [comment]
Yes. Babel is also wrong on this, only esbuild is right. However esbuild's solution would add quite a bit of complexity to implementation.

**Comment by magic-akari** [comment]
The order of execution in the deconstruction is incorrect, things are more complex than I anticipated.
